server/src/test/java/org/elasticsearch/cluster/IndexingWithMappingUpdateIT.java

```diff
@@ -0,0 +1,193 @@
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+import org.elasticsearch.action.bulk.BulkAction;
+import org.elasticsearch.action.index.IndexRequestBuilder;
+import org.elasticsearch.action.index.IndexResponse;
+import org.elasticsearch.client.Client;
+import org.elasticsearch.common.settings.Settings;
+import org.elasticsearch.common.xcontent.XContentType;
+import org.elasticsearch.discovery.zen.PublishClusterStateAction;
+import org.elasticsearch.plugins.Plugin;
+import org.elasticsearch.test.ESIntegTestCase;
+import org.elasticsearch.test.ESIntegTestCase.ClusterScope;
+import org.elasticsearch.test.InternalTestCluster;
+import org.elasticsearch.test.transport.MockTransportService;
+import org.elasticsearch.transport.TransportService;
+import org.junit.Before;
+
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.List;
+import java.util.concurrent.CopyOnWriteArrayList;
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.stream.Collectors;
+
+import static org.elasticsearch.client.Requests.createIndexRequest;
+import static org.elasticsearch.test.hamcrest.ElasticsearchAssertions.assertAcked;
+import static org.hamcrest.Matchers.hasSize;
+import static org.hamcrest.Matchers.is;
+
+@ClusterScope(scope = ESIntegTestCase.Scope.SUITE, numDataNodes = 0)
+public class IndexingWithMappingUpdateIT extends ESIntegTestCase {
+    private String index = "test";
+    private String type = "type1";
+    private String masterName;
+    private String primaryNodeName;
+    private String replicaNodeName;
+    private Client client;
+    private InternalTestCluster cluster;
+
+    @Override
+    protected Collection<Class<? extends Plugin>> nodePlugins() {
+        return Arrays.asList(MockTransportService.TestPlugin.class);
+    }
+
+    @Before
+    public void layout() throws Exception {
+        // dedicated master + 1 primary + 1 replica
+        cluster = internalCluster();
+        masterName = cluster.startMasterOnlyNode();
+        primaryNodeName = cluster.startDataOnlyNode();
+
+        assertAcked(admin().indices()
+            .create(createIndexRequest(index)
+                .settings(Settings.builder()
+                    .put("index.number_of_replicas", 1)
+                    .put("index.number_of_shards", 1)))
+            .get());
+
+
+        client = cluster.client(primaryNodeName);
+
+        // create index in advance to have allocate primary
+        client.prepareIndex(index, type, "d").setSource("{ }", XContentType.JSON).get();
```

Comment:
I'm confused by this. It doesn't create the index as indicated by the comment. Maybe it is here to force the primary to be fully allocated before starting the replica? If so, I think it'd be better to do that more explicitly, with something like the following.

```
assertBusy(() -> {
            ClusterState clusterState = client().admin().cluster().prepareState().get().getState();
            String nodeId = clusterState.getRoutingTable().index(index).shard(0).primaryShard().currentNodeId();
            assertThat(clusterState.getRoutingNodes().node(nodeId).node().getName(), is(primaryNodeName));
        });
```

Alternatively we could just start both data nodes, create the index, and then ask it which one is the primary and which one is the replica.